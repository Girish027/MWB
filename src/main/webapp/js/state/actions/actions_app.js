
import _ from 'lodash';
import getUrl, { pathKey } from 'utils/apiUrls';
import { fetchUtil } from 'utils/fetchUtils';
import goToRoute from 'utils/routeHelpers';
import Constants from 'constants/Constants';
import AmplitudeConstants from 'constants/AmplitudeConstants';
import { logAmplitudeEvent } from 'utils/amplitudeUtils';
import axios from 'axios';
import api from 'utils/api';
import { setDefaultsForAPI } from 'utils/api';
import * as types from './types';

// Note to developers:
// Both Warning and Good Messages are shown for the period of notificationInterval.
// Bad, Persistant - till user closes the notification. This should only be used sparingly.
export const callLogIngestAPI = (message, logLevel) => {
  const promise = new Promise((resolve, reject) => {
    const url = getUrl(pathKey.logIngest, { logLevel });
    return new Promise((resolve, reject) => {
      api.post({
        url,
        headers: {
          'Content-Type': 'text/plain',
        },
        data: message,
      });
    });
  });
};

export const displayBadRequestMessage = (message) => (dispatch) => {
  dispatch({
    type: types.BAD_REQUEST,
    value: message,
  });

  callLogIngestAPI(message, Constants.LOG_LEVEL_ERROR);
};

export const displayWarningRequestMessage = (message, notificationInterval = Constants.NOTIFICATION.interval) => (dispatch) => {
  dispatch({
    type: types.WARNING_REQUEST,
    value: message,
  });

  callLogIngestAPI(message, Constants.LOG_LEVEL_WARNING);

  setTimeout(() => {
    dispatch(stopShowingServerMessage());
  }, notificationInterval);
};

export const displayGoodRequestMessage = (message, notificationInterval = Constants.NOTIFICATION.interval) => (dispatch) => {
  dispatch({
    type: types.GOOD_REQUEST,
    value: message,
  });

  setTimeout(() => {
    dispatch(stopShowingServerMessage());
  }, notificationInterval);
};

export const displayPersistantMessage = message => ({
  type: types.PERSISTENT_REQUEST,
  value: message,
});

export const stopShowingServerMessage = () => ({
  type: types.APP_STOP_SHOWING_SERVER_MESSAGE,
});

export const updateLoginSuccess = csrfToken => ({
  type: types.LOGIN_SUCCESS,
  value: csrfToken,
});

export const updateLoginInfo = (userId, userName, userDetails) => ({
  type: types.LOGIN_INFO,
  value: {
    userId,
    userName,
    userDetails,
  },
});

export const logoutFromOktaSuccess = () => ({
  type: types.LOGOUT_OKTA,
});

export const logoutFromMWBSuccess = () => ({
  type: types.LOGOUT_MWB,
});

export const contextMenuChange = contextMenuState => ({
  type: types.CONTEXT_MENU_CHANGE,
  contextMenuState,
});

export const modalDialogChange = modalDialogState => ({
  type: types.MODAL_DIALOG_CHANGE,
  modalDialogState,
});

const actionFileRemoveRequest = ({ fileId }) => ({
  type: types.FILE_REMOVE_REQUEST,
  fileId,
});

const actionFileRemoveError = ({ fileId, error }) => ({
  type: types.FILE_REMOVE_ERROR,
  fileId,
  error,
});

const actionFileRemoveSuccess = ({ fileId }) => ({
  type: types.FILE_REMOVE_SUCCESS,
  fileId,
});

export const updateUserGroups = data => ({
  type: types.UPDATE_USER_GROUPS,
  data,
});

const onFetchFailureRemoveFile = (error, params) => (dispatch) => {
  const { fileId } = params;
  dispatch(actionFileRemoveError({ fileId, error }));
};

const onFetchSuccessRemoveFile = (data, params) => (dispatch) => {
  const { fileId } = params;
  dispatch(actionFileRemoveSuccess({ fileId }));
};

export const removeFile = ({ fileId }) => (dispatch, getState) => {
  dispatch(actionFileRemoveRequest({ fileId }));

  const state = getState();
  const fetchUrl = getUrl(pathKey.files, { fileId, clientId: state.header.client.id });

  fetchUtil({
    fetchUrl,
    fetchMethod: 'delete',
    onFetchSuccess: onFetchSuccessRemoveFile,
    onFetchFailure: onFetchFailureRemoveFile,
    dispatch,
    getState,
    params: {
      fileId,
    },
  });
};

const onFetchSuccessGetUserGroups = data => (dispatch) => {
  dispatch(updateUserGroups(data));
};

export const getUserGroups = ({ csrfToken }) => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.userGroups, { csrfToken });

  fetchUtil({
    fetchUrl,
    fetchMethod: 'get',
    onFetchSuccess: onFetchSuccessGetUserGroups,
    dispatch,
    getState,
  });
};

export const onFetchSuccessGetCSRFToken = data => (dispatch, getState) => {
  const csrfToken = data.value;
  setDefaultsForAPI(csrfToken);
  dispatch(updateLoginSuccess(csrfToken));
  dispatch(getUserGroups({ csrfToken }));
};

export const getCSRFToken = userId => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.csrfToken);

  fetchUtil({
    fetchUrl,
    fetchMethod: 'get',
    onFetchSuccess: onFetchSuccessGetCSRFToken,
    dispatch,
    getState,
    params: {
      userId,
    },
  });
};

const logoutFromMWB = () => (dispatch, getState) => {
  dispatch(logoutFromOktaSuccess());
  dispatch(logoutFromMWBSuccess());
};

export const logoutFromOkta = () => (dispatch, getState) => {
  const state = getState();
  const { oAuthLogoutURL } = state.app;
  if (oAuthLogoutURL) {
    fetch(oAuthLogoutURL, {
      method: 'DELETE',
      credentials: 'include',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    })
      .then((response) => {
        window.location.href = Constants.LOGOUT;
        switch (response.status) {
        case 204:
          logoutFromMWB();
          logAmplitudeEvent(AmplitudeConstants.LOGOUT_EVENT, state);
          break;
        case 404:
        default:
          // TODO send log for these scenarios
          break;
        }
      })
      .catch((error) => {
        // TODO send log for these scenarios
      });
  }
};

export const changeRoute = (routeName, data, history) => (dispatch, getState) => {
  const state = getState();
  const routeData = Object.assign({}, {
    client: !_.isNil(state.header.client) ? state.header.client : {
      id: 0,
      name: 'Loading...',
    },
    projectId: state.selectedProjectId,
  }, data);
  goToRoute(routeName, routeData, history);
};
