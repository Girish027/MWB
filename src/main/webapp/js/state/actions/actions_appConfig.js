import 'regenerator-runtime/runtime';

import getUrl, { pathKey } from 'utils/apiUrls';

import { fetchUtil } from 'utils/fetchUtils';
import AmplitudeConstants from 'constants/AmplitudeConstants';
import { initializeAnalytics, logAmplitudeEvent } from 'utils/amplitudeUtils';
import api from 'utils/api';
import * as types from './types';

const appConfigSuccess = data => ({
  type: types.APP_CONFIG_SUCCESS,
  data,
});

const appConfigFailure = data => ({
  type: types.APP_CONFIG_FAILURE,
  data,
});

const appConfigRequest = () => ({
  type: types.APP_CONFIG_REQUEST,
});

export const onAppConfigSuccess = data => (dispatch, getState) => {
  dispatch(appConfigSuccess(data));
  const state = getState();
  initializeAnalytics(state);
  logAmplitudeEvent(AmplitudeConstants.LOGIN_EVENT, state);
};

export const onAppConfigFailure = error => (dispatch) => {
  dispatch(appConfigFailure(error));
};

export const fetchAppConfig = () => (dispatch, getState) => {
  dispatch(appConfigRequest());
  const fetchUrl = getUrl(pathKey.appConfig);

  api.get({
    url: fetchUrl,
    onApiSuccess: onAppConfigSuccess,
    onApiError: onAppConfigFailure,
    dispatch,
    getState,
  });
};
