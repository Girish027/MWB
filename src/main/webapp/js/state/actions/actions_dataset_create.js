import fetch from 'isomorphic-fetch';
import errorMessageUtil from 'utils/ErrorMessageUtil';
import { addDatasetToProject, fetchDatasetTransform } from 'state/actions/actions_datasets';
import * as apiUtils from 'utils/apiUtils';
import getUrl, { pathKey } from 'utils/apiUrls';
import { getLanguage } from 'state/constants/getLanguage';
import { logAmplitudeEvent } from 'utils/amplitudeUtils';
import AmplitudeConstants from 'constants/AmplitudeConstants';
import Constants from 'constants/Constants';
import api from 'utils/api';
import * as projectsActions from 'state/actions/actions_projects';
import { escapeRegExp } from 'utils/StringUtils';
import * as appActions from './actions_app';
import * as types from './types';
import { colors } from '../../styles';

const languageConstants = getLanguage();
const { DISPLAY_MESSAGES } = languageConstants;

export const reset = () => ({
  type: types.CREATE_DATASET_DIALOG_RESET,
});

export const fileUploadSuccess = ({ token, columns, previewData }) => ({
  type: types.CREATE_DATASET_DIALOG_FILE_UPLOAD_SUCCESS,
  token,
  columns,
  previewData,
});

export const changeStep = ({ step }) => ({
  type: types.CREATE_DATASET_DIALOG_CHANGE_STEP,
  step,
});

export const columnsBind = ({
  bindingArray, isBindingValid, columnsBinding, isPreSelected,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND,
  bindingArray,
  isBindingValid,
  columnsBinding,
  isPreSelected,
});

export const changeAutoTag = ({ autoTag }) => ({
  type: types.CREATE_DATASET_DIALOG_AUTO_TAG_CHANGE,
  autoTag,
});

export const changesStartTransform = ({ startTransform }) => ({
  type: types.CREATE_DATASET_DIALOG_START_TRANSFORM_CHANGE,
  startTransform,
});

export const changeFirstRowSkip = ({ skipFirstRow }) => ({
  type: types.CREATE_DATASET_DIALOG_FISRT_ROW_SKIP_CHANGE,
  skipFirstRow,
});

const columnsBindRequest = ({ token, bindingArray, skipFirstRow }) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_REQUEST,
  token,
  bindingArray,
  skipFirstRow,
});

const columnsBindError = ({
  token, bindingArray, skipFirstRow, error,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_ERROR,
  token,
  bindingArray,
  skipFirstRow,
  error,
});

const columnsBindSuccess = ({
  token, bindingArray, skipFirstRow, fileId,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_SUCCESS,
  token,
  bindingArray,
  skipFirstRow,
  fileId,
});

export const requestColumnsBind = ({ token, bindingArray, skipFirstRow }) => (dispatch, getState) => new Promise((resolve, reject) => {
  dispatch(columnsBindRequest({ token, bindingArray, skipFirstRow }));

  const state = getState();
  const { csrfToken } = state.app;
  const clientId = state.header.client.id;

  const fetchUrl = getUrl(pathKey.importDatasetColumnsBind, { token, skipFirstRow, clientId });

  fetch(fetchUrl, {
    credentials: 'same-origin',
    method: 'POST',
    headers: {
      'X-CSRF-TOKEN': csrfToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(bindingArray),
  })
    .then(errorMessageUtil.handleErrors)
    .then(response => response.json())
    .then((response) => {
      /* fileId, name, systemName */
      const bindData = Object.assign({}, {
        token,
        bindingArray,
        skipFirstRow,
        fileId: response.fileId,
      });
      dispatch(columnsBindSuccess(bindData));
      resolve(bindData);
    })
    .catch((error) => {
      dispatch(columnsBindError({
        token, bindingArray, skipFirstRow, error,
      }));
      dispatch(appActions.modalDialogChange(null));
      errorMessageUtil.dispatchError(error, dispatch, appActions.displayBadRequestMessage);
      reject(error);
      dispatch(reset());
    });
});

const createDatasetRequest = () => ({
  type: types.CREATE_DATASET_REQUEST,
});

const createDatasetError = () => ({
  type: types.CREATE_DATASET_ERROR,
});

const createDatasetSuccess = () => ({
  type: types.CREATE_DATASET_SUCCESS,
});

export const createDataset = ({
  projectId, clientId, dataType, locale, name, uri, description, autoTagDataset, mapToProjectId,
}) => (dispatch, getState) => new Promise((resolve, reject) => {
  dispatch(createDatasetRequest());

  const state = getState();
  const { userDetails = Constants.OBJECT_DEFAULT_VALUE, csrfToken } = state.app;
  const { userType = Constants.DEFAULT_VALUE } = userDetails;
  const fetchUrl = getUrl(pathKey.datasets, { clientId });
  let source = Constants.DATASET_EXTERNAL_FLAG;

  logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_EVENT, state, {
    projectId, clientDBId: clientId,
  });

  if (userType == Constants.USER_TYPE_INTERNAL) {
    source = Constants.DATASET_INTERNAL_FLAG;
  }

  fetch(fetchUrl, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'X-CSRF-TOKEN': csrfToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      dataset: {
        projectId, clientId, dataType, locale, name, source, uri, description,
      },
      projectId,
      autoTagDataset,
    }),
  })
    .then(errorMessageUtil.handleErrors)
    .then(response => response.json())
    .then((response) => {
      const newDataset = apiUtils.normalizeIds(response);
      const dataset = Object.assign({ status: 'NULL', type: newDataset.dataType, description: newDataset.description || '' }, newDataset);
      dispatch(createDatasetSuccess());
      if (!mapToProjectId) {
        resolve(dataset);
      } else {
        dispatch(addDatasetToProject({
          datasetId: dataset.id,
          projectId,
          dataset,
        }))
          .then(() => {
            logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_COMPLETED_EVENT, state, {
              projectId, datasetId: dataset.id, clientDBId: clientId,
            });
            dispatch(projectsActions.refreshProjectsByClient(clientId));
            resolve(dataset);
          })
          .catch((error) => {
            logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
              projectId, clientDBId: clientId,
            });
            dispatch(appActions.modalDialogChange(null));
            reject(error);
          });
      }
    })
    .catch((error) => {
      dispatch(appActions.modalDialogChange(null));
      switch (error.status) {
      case 400:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.createFailed,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.createFailed));
        break;
      case 403:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetOperationNotAllowed,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.datasetOperationNotAllowed));
        break;
      case 409:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetAlreadyExists(name),
        });
        dispatch(appActions.displayWarningRequestMessage(DISPLAY_MESSAGES.datasetAlreadyExists(name)));
        break;
      default:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetCreateError,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.datasetCreateError));
        errorMessageUtil.dispatchError(error, dispatch, appActions.displayBadRequestMessage);
      }
      dispatch(createDatasetError());
      reject(error);
    });
});

export const onDatasetImportSuccess = (response = {}, importData) => (dispatch, getState) => {
  let { token, columns, previewData } = response;
  if (previewData && previewData.length > 10) {
    previewData = previewData.slice(0, 10);
  }
  dispatch(fileUploadSuccess({ token, columns, previewData }));
  dispatch(changeStep({ step: 'mapping' }));
  dispatch(columnBinding(importData));
};

export const onDatasetImportFailure = (error = {}) => (dispatch, getState) => {
  const { code } = error;
  if (code == 400) {
    const { message } = error;
    dispatch(appActions.displayBadRequestMessage(message));
  } else if (code == 401) {
    dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.sessionExpired));
  } else {
    dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.fileParseFail));
  }
  dispatch(reset());
  dispatch(appActions.modalDialogChange(null));
};

export const importFile = importData => (dispatch, getState) => {
  const { dropZone } = importData;
  const url = getUrl(pathKey.importDataset);
  const formData = new FormData();
  if (dropZone) {
    formData.append('datatype', null);
    formData.append(Constants.DATASET_FILE, dropZone, dropZone.name);
  }

  return new Promise((resolve, reject) => {
    api.post({
      dispatch,
      getState,
      url,
      data: formData,
      onApiSuccess: response => onDatasetImportSuccess(response, importData),
      onApiError: onDatasetImportFailure,
    });
  });
};

export const columnBinding = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog } = state;
  const {
    columns, previewData, token,
    isMappingRequestLoading,
  } = createDatasetDialog;
  const { isSomeColumnsPreSelected, columnsBinding } = tryPreSelectColumns(columns,
    previewData, state.createDatasetDialog.columnsBinding);

  let requiredColumns = [];
  let columnsByName = {};

  columns.forEach(column => {
    columnsByName[column.name] = column;
    if (column.required || column.name === 'rutag') {
      requiredColumns.push(column.name);
    }
  });

  if (isSomeColumnsPreSelected) {
    let isBindingValid = true;
    requiredColumns.forEach(column => {
      if (typeof columnsBinding[column] === 'undefined') {
        isBindingValid = false;
      }
    });

    const bindingArray = [];
    for (const columnName in columnsBinding) {
      if (!columnsBinding.hasOwnProperty(columnName)) {
        continue;
      }
      const column = columnsByName[columnName];
      bindingArray.push({
        id: `${column.id}`,
        columnName: column.name,
        columnIndex: `${columnsBinding[columnName]}`,
        displayName: column.displayName,
      });
    }

    isBindingValid = isBindingValid && bindingArray.length > 0;

    dispatch(columnsBind({
      bindingArray, isBindingValid, columnsBinding, isPreSelected: true,
    }));

    if (isBindingValid && bindingArray.length > 0 && token && !isMappingRequestLoading) {
      dispatch(confirmDatasetUpload(importData));
    } else {
      dispatch(showDatasetMapperDialog(importData));
    }
  } else {
    dispatch(showDatasetMapperDialog(importData));
  }
};

const showDatasetMapperDialog = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog } = state;
  dispatch(appActions.modalDialogChange({
    dispatch,
    createDatasetDialog,
    type: Constants.DIALOGS.DATASET_COLUMN_MAPPING_DIALOG,
    onOk: () => dispatch(confirmDatasetUpload(importData)),
  }));
};

const confirmDatasetUpload = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog, app } = state;
  const { userId, csrfToken } = app;
  const { token, bindingArray, skipFirstRow } = createDatasetDialog;
  const {
    name, description, projectId,
    clientId, locale, dataType,
  } = importData;
  dispatch(appActions.modalDialogChange({
    dispatch,
    type: Constants.DIALOGS.PROGRESS_DIALOG,
    message: Constants.UPLOAD_DATASET_PROGRESS,
    header: Constants.UPLOADING_DATASET,
    showHeader: true,
    showFooter: true,
    cancelVisible: true,
    closeIconVisible: true,
    cancelChildren: Constants.CLOSE,
    styleOverride: {
      childContainer: {
        marginTop: '30px',
        marginBottom: '10px',
      },
      content: {
        top: '160px',
        maxWidth: '444px',
        maxHeight: '333px',
        left: 'calc((100vw - 500px) / 2)',
      },
      cancel: {
        color: `${colors.white}`,
        backgroundColor: `${colors.cobalt}`,
        paddingLeft: '25px',
        paddingRight: '25px',
        ':hover': {
          color: `${colors.white}`,
          backgroundColor: `${colors.prussianBlue}`,
        },
      },
    },
  }));
  dispatch(requestColumnsBind({
    token, bindingArray, skipFirstRow,
  })).then((bindData) => {
    const { fileId } = bindData;
    dispatch(createDataset({
      projectId,
      clientId,
      dataType,
      locale,
      name,
      uri: `https://tagging.247-inc.com:8443/nltools/private/v1/files/${fileId}`,
      description,
      autoTagDataset: true,
      mapToProjectId: projectId,
    })).then((dataset) => {
      setTimeout(() => {
        dispatch(fetchDatasetTransform(userId, dataset.id, clientId, projectId, csrfToken, false));
      }, 2000);
      dispatch(appActions.displayGoodRequestMessage(DISPLAY_MESSAGES.datasetCreated(dataset.name)));
      dispatch(reset());
    });
  });
};

const tryPreSelectColumns = (columns, data, columnsBindingInput) => {
  const firstRow = data.length ? data[0] : [];
  const columnsBinding = Object.assign({}, columnsBindingInput);
  if (!firstRow.length || !columns.length || (Object.keys(columnsBinding)).length != 0) {
    return { isSomeColumnsPreSelected: false, columnsBinding };
  }

  columns.forEach((c) => {
    let foundIndexes = [];
    let usedIndexes = {};

    // pass 1 - exact match
    let tests = [
      new RegExp(`^${escapeRegExp(c.name)}$`, 'i'),
      new RegExp(`^${escapeRegExp(c.displayName)}$`, 'i'),
    ];

    firstRow.forEach((newData1, index) => {
      for (let i = 0; i < tests.length; i++) {
        if (tests[i].test(newData1.trim())) {
          foundIndexes.push(index);
          break;
        }
      }
    });

    if (!foundIndexes.length) {
      // pass 2 - partial match
      tests = [
        new RegExp(escapeRegExp(c.name), 'i'),
        new RegExp(escapeRegExp(c.displayName), 'i'),
      ];

      firstRow.forEach((newData2, index) => {
        for (let i = 0; i < tests.length; i++) {
          if (tests[i].test(newData2)) {
            foundIndexes.push(index);
            break;
          }
        }
      });
    }

    if (foundIndexes.length === 1 && !usedIndexes[foundIndexes[0]]) {
      columnsBinding[c.name] = foundIndexes[0];
      usedIndexes[foundIndexes[0]] = true;
    }
  });
  return { isSomeColumnsPreSelected: (Object.keys(columnsBinding)).length != 0, columnsBinding };
};

export const fileUpload = (file, handleSuccess, handleError) => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.importDataset);
  const formData = new FormData();
  formData.append('datatype', null);
  formData.append('file', file);
  let req = new XMLHttpRequest();
  req.open('post', fetchUrl, true);
  req.onload = (event) => {
    let { response } = req;
    if (req.status == 200) {
      handleSuccess(response);
    } else {
      handleError(response);
    }
  };
  req.send(formData);
};
