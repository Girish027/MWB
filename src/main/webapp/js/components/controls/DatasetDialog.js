import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Dialog } from '@tfs/ui-components';
import { importFile } from 'state/actions/actions_dataset_create';
import isFeatureEnabled, { featureFlagDefinitions } from 'utils/FeatureFlags';
import Constants from '../../constants/Constants';
import Form from '../Form/Form';
import * as appActions from '../../state/actions/actions_app';
import { pathKey } from '../../utils/apiUrls';

import createDatasetUiSchema from '../schema/createDataset/uiSchema.json';
import createDatasetUiSchemaExt from '../schema/createDataset/uiSchemaExternal.json';
import createDatasetJsonSchema from '../schema/createDataset/jsonSchema.json';
import createDatasetJsonSchemaExt from '../schema/createDataset/jsonSchemaExternal.json';
import CustomTextField from '../fields/CustomTextField';
import CustomDropDown from '../fields/CustomDropDown';
import CustomDropZone from '../fields/CustomDropZone';

export default class DatasetDialog extends Component {
  constructor(props) {
    super(props);
    this.form = React.createRef();
    this.props = props;

    this.onClickNext = this.onClickNext.bind(this);
    this.onClickCancel = this.onClickCancel.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onFormChange = this.onFormChange.bind(this);

    this.styleOverride = Object.assign({}, {
      childContainer: {
        marginTop: '0px',
        marginBottom: '0px',
      },
      container: {
        width: '446px',
        height: '460px',
        display: 'grid',
        gridTemplateRows: '60px auto 60px',
      },
    },
    this.props.styleOverride);

    this.state = {
      formData: {},
      isValid: false,
      datasetFile: null,
    };

    this.fields = {
      text: CustomTextField,
      select: (props, url = pathKey.datatypes) => <CustomDropDown {...props} url={pathKey.datatypes} defaultValue={Constants.CHOOSE_TYPE} />,
      dropZone: (props) => <CustomDropZone {...props} />,
    };
  }

  onClickNext() {
    this.form.current.submit();
  }

  onClickCancel() {
    const { dispatch } = this.props;
    dispatch(appActions.modalDialogChange(null));
  }

  onSubmit() {
    const { formData } = this.state;
    const {
      dispatch, project, userFeatureConfiguration,
    } = this.props;
    const { id: projectId, clientId, locale } = project;

    const { names } = featureFlagDefinitions;

    let importData = {
      ...formData,
      projectId,
      clientId,
      locale,
    };

    if (!isFeatureEnabled(names.datasetType, userFeatureConfiguration)) {
      importData = { ...importData, dataType: 'Chat/Text' };
    }

    dispatch(importFile(importData));
  }

  onFormChange(data, errors) {
    if (errors.length === 0 && Object.entries(data.dropZone).length !== 0) {
      this.setState({
        formData: data,
        isValid: true,
      });
    } else if (errors.length > 0) {
      this.setState({
        isValid: false,
      });
    }
  }

  render() {
    const {
      header, size, formData, errorType, userFeatureConfiguration, ...otherProps
    } = this.props;

    const { isValid } = this.state;
    const { names } = featureFlagDefinitions;

    return (
      <div>
        <Dialog
          size={size}
          headerChildren={header}
          {...otherProps}
          isOpen
          showBodySeperator={false}
          okVisible
          okDisabled={!isValid}
          okChildren={Constants.NEXT}
          cancelVisible
          cancelChildren={Constants.CANCEL}
          closeIconVisible
          styleOverride={this.styleOverride}
          onClickOk={this.onClickNext}
          onClickCancel={this.onClickCancel}
          onClickClose={this.onClickCancel}
          centerContent={false}
        >
          <div>
            <div>
              <Form
                uiSchema={isFeatureEnabled(names.datasetType, userFeatureConfiguration) ? createDatasetUiSchema : createDatasetUiSchemaExt}
                jsonSchema={isFeatureEnabled(names.datasetType, userFeatureConfiguration) ? createDatasetJsonSchema : createDatasetJsonSchemaExt}
                fields={this.fields}
                onSubmit={this.onSubmit}
                onChange={this.onFormChange}
                formData={formData}
                ref={this.form}
                liveValidate
              />
            </div>
          </div>
        </Dialog>
      </div>
    );
  }
}

DatasetDialog.propTypes = {
  dispatch: PropTypes.func,
  project: PropTypes.object,
  header: PropTypes.string,
  closeIconVisible: PropTypes.bool,
  onClickOk: PropTypes.func,
  onClickClose: PropTypes.func,
  styleOverride: PropTypes.object,
  size: PropTypes.string,
  formData: PropTypes.object,
  errorType: PropTypes.string,
  userFeatureConfiguration: PropTypes.object,
};

DatasetDialog.defaultProps = {
  dispatch: () => {},
  project: {},
  closeIconVisible: true,
  onClickOk: () => {},
  onClickClose: () => {},
  styleOverride: {},
  size: 'large',
  formData: {},
  errorType: '',
  header: Constants.CREATE_MODEL,
  userFeatureConfiguration: {},
};
