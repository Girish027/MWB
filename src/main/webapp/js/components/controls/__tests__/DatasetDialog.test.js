import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import * as actionsApp from 'state/actions/actions_app';
import Constants from 'constants/Constants';
import * as actionsDatasetCreate from 'state/actions/actions_dataset_create';
import DatasetDialog from 'components/controls/DatasetDialog';

describe('<DatasetDialog />', () => {
  let wrapper;
  const props = {
    header: Constants.UPLOAD_DATASET,
    onCancel: () => {},
    dispatch: () => {},
  };

  const getShallowWrapper = (propsObj) => shallow(<DatasetDialog
    {...props}
  />);

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance with test props', () => {
    beforeEach(() => {
      wrapper = getShallowWrapper(props);
    });

    test('should exist', () => {
      expect(wrapper.exists()).toBe(true);
    });

    test('Snapshots should render correctly', () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Function Call Handler:', () => {
    beforeAll(() => {
      props.header = Constants.CREATE_MODEL,
      props.dispatch = jest.fn();
      props.onCancel = jest.fn();
      actionsApp.modalDialogChange = jest.fn(() => 'called modalDialogChange');
      actionsDatasetCreate.importFile = jest.fn(() => 'called importFile');
    });

    beforeEach(() => {
      wrapper = shallow(<DatasetDialog
        {...props}
      />);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('onClickCancel:', () => {
      test('should onClickCancel call click handler', () => {
        wrapper.instance().onClickCancel();
        expect(props.dispatch).toHaveBeenCalledWith('called modalDialogChange');
        expect(actionsApp.modalDialogChange).toHaveBeenCalledWith(null);
      });
    });

    describe('onSubmit:', () => {
      test('should call importFile action', () => {
        wrapper.instance().onSubmit();
        expect(props.dispatch).toHaveBeenCalledWith('called importFile');
      });
    });
  });

  describe('onFormChange:', () => {
    test('should change the state when there are no error', () => {
      const data = {
        name: 'abc',
        description: 'form',
        dropZone: [{ acceptedFile: 1 }],
      };
      wrapper.instance().onFormChange(data, []);
      expect(wrapper.state().formData).toBe(data);
      expect(wrapper.state().isValid).toBe(true);
    });

    test('should change the state when there are no error', () => {
      const data = {
        name: 'abc',
        description: 'form',
      };
      wrapper.instance().onFormChange(data, [{ name: 'is required' }]);
      expect(wrapper.state().isValid).toBe(false);
    });
  });
});
