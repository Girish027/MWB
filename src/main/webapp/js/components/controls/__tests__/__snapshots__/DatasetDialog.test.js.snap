// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatasetDialog /> Creating an instance with test props Snapshots should render correctly 1`] = `
<div>
  <withTheme(DialogComponent)
    cancelChildren="CANCEL"
    cancelVisible={true}
    centerContent={false}
    closeIconVisible={true}
    dispatch={[Function]}
    headerChildren="Upload Dataset"
    isOpen={true}
    okChildren="NEXT"
    okDisabled={true}
    okVisible={true}
    onCancel={[Function]}
    onClickCancel={[Function]}
    onClickClose={[Function]}
    onClickOk={[Function]}
    project={Object {}}
    showBodySeperator={false}
    size="large"
    styleOverride={
      Object {
        "childContainer": Object {
          "marginBottom": "0px",
          "marginTop": "0px",
        },
        "container": Object {
          "display": "grid",
          "gridTemplateRows": "60px auto 60px",
          "height": "460px",
          "width": "446px",
        },
      }
    }
  >
    <div>
      <div>
        <Form
          fields={
            Object {
              "dropZone": [Function],
              "select": [Function],
              "text": [Function],
            }
          }
          formContext={Object {}}
          formData={Object {}}
          jsonSchema={
            Object {
              "properties": Object {
                "description": Object {
                  "optional": true,
                  "title": "Description",
                  "type": "string",
                },
                "dropZone": Object {
                  "minLength": 1,
                  "type": "object",
                },
                "name": Object {
                  "isRequired": true,
                  "title": "Name",
                  "type": "string",
                },
              },
              "required": Array [
                "name",
                "dropZone",
              ],
              "title": "",
              "type": "object",
            }
          }
          liveValidate={true}
          onChange={[Function]}
          onSubmit={[Function]}
          showErrorList={false}
          uiSchema={
            Object {
              "description": Object {
                "ui:field": "text",
              },
              "dropZone": Object {
                "ui:field": "dropZone",
              },
              "name": Object {
                "ui:field": "text",
              },
            }
          }
        />
      </div>
    </div>
  </withTheme(DialogComponent)>
</div>
`;
